### PR TITLE
Throw a parsing exception when boost is set in span_or query (#28390)

### PR DESCRIPTION
**Pull request by cbismuth**
This pull request **is not complete** will contain guards against setting boost values in compound span queries as they are ignored in Lucene.

Before going futher I'd like to have a validation from reviewers about suggested implementation.

For more details, please see issue https://github.com/elastic/elasticsearch/issues/28390.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java [hunk 1]
```diff
@@ -49,7 +50,7 @@ public SpanOrQueryBuilder(SpanQueryBuilder initialClause) {
         if (initialClause == null) {
             throw new IllegalArgumentException("[" + NAME + "] must include at least one clause");
         }
-        clauses.add(initialClause);
+        addClause(initialClause);
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java]: I think this is not necessary. The only place where we should do the check is inside `fromXContent` method for all span query builders : `SpanOrQueryBuilder`, `SpanTermQueryBuilder` etc.

File: server/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java [hunk 2]
```diff
@@ -58,7 +59,7 @@ public SpanOrQueryBuilder(SpanQueryBuilder initialClause) {
     public SpanOrQueryBuilder(StreamInput in) throws IOException {
         super(in);
         for (QueryBuilder clause: readQueries(in)) {
-            clauses.add((SpanQueryBuilder) clause);
+            addClause((SpanQueryBuilder) clause);
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java]: I think this is not necessary.  This comes from another node that presumably already failed a span query if it contains boost.
The only place where we should do the check is inside `fromXContent` method for all span query builders : `SpanOrQueryBuilder`, `SpanTermQueryBuilder` etc

File: server/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java [hunk 3]
```diff
@@ -74,10 +75,23 @@ public SpanOrQueryBuilder addClause(SpanQueryBuilder clause) {
         if (clause == null) {
             throw new IllegalArgumentException("[" + NAME + "] inner clause cannot be null");
         }
+        checkBoostValue(clause);
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java]: Similar comment here: I think this is not necessary. The only place where we should do the check is inside fromXContent method for all span query builders : SpanOrQueryBuilder, SpanTermQueryBuilder etc.

File: server/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java [hunk 4]
```diff
@@ -115,14 +129,16 @@ public static SpanOrQueryBuilder fromXContent(XContentParser parser) throws IOEx
                         if (query instanceof SpanQueryBuilder == false) {
                             throw new ParsingException(parser.getTokenLocation(), "spanOr [clauses] must be of type span query");
                         }
-                        clauses.add((SpanQueryBuilder) query);
+                        final SpanQueryBuilder clause = (SpanQueryBuilder) query;
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java]: I think these 3 lines are not necessary, as each inner SpanQueryBuilder within their `fromXContent` should check for `boost` and fail if `boost` provided.

File: server/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java [hunk 5]
```diff
@@ -115,14 +129,16 @@ public static SpanOrQueryBuilder fromXContent(XContentParser parser) throws IOEx
...
+                        checkBoostValue(parser.getTokenLocation(), clause);
+                        clauses.add(clause);
                     }
                 } else {
                     throw new ParsingException(parser.getTokenLocation(), "[span_or] query does not support [" + currentFieldName + "]");
                 }
             } else {
                 if (AbstractQueryBuilder.BOOST_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
-                    boost = parser.floatValue();
+                    throw new ParsingException(parser.getTokenLocation(), "spanOr [clauses] can't have boost value");
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java]: That is great and I think the only necessary change in `SpanOrQueryBuilder`.  All other SpanQueryBuilders should have similar line in their `fromXContent` methods.

**Review by mayya-sharipova — CHANGES_REQUESTED** [review]
@cbismuth  Thanks for tacking this problem.  Good first attempt! I have left some comments how we may change the code.

**Comment by cbismuth** [comment]
Thanks a lot for your review @mayya-sharipova, I'll update my PR and add tests.

**Comment by cbismuth** [comment]
Hi @mayya-sharipova, clauses of compound span queries aren't aware that `boost` field isn't supported. Therefore, when converting a `span_or` query to JSON, each clause emits a `boost` field (with `1.0` as default).

I'm not sure serialization should be changed, but to fix the reported issue I'd suggest to throw a `ParsingException` only when `boost` value isn't equal to default `boost` value (i.e. `1.0`). So that JSON serialization will keep on working as it does today and parsing will throw an exception when `boost` has been tuned.

I've added a commit to this PR to show you the changes I'd like to make. Could you please tell me what do you think of this? Thank you.

**Comment by mayya-sharipova** [comment]
@cbismuth  Your strategy of checking for `DEFAULT_BOOST` and throwing an exception only if different from `DEFAULT_BOOST`, makes sense to me.   Thanks for noticing it!

**Comment by cbismuth** [comment]
Thank you for your review @mayya-sharipova :+1: I think I have enough information to implement a complete patch.

**Comment by cbismuth** [comment]
Hi @mayya-sharipova, PR is complete and ready for review. I stay available for any further change, thank you.

File: server/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java (hunk 4 above)
Comment (cbismuth) [reply server/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java]: Compound `span_or` queries can contain `span_term` inner queries and setting `boost` in `span_term` queries is allowed, that's why I've added this check here, do you agree with it? See this [test](https://github.com/elastic/elasticsearch/pull/34112/files#diff-9103f862a98e22ff9bc652661ba0a192R110).

**Comment by cbismuth** [comment]
Hi, this PR is synced with latest `master` and ready for review, thanks.

**Comment by cbismuth** [comment]
Hi, I have no doubt you're quite busy on 7.x release, but it would be totally awesome if this PR could be reviewed, thanks a lot :+1:

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

**Comment by mayya-sharipova** [comment]
@cbismuth  I am very sorry, this PR somehow slipped my review. I will check it tomorrow after tests pass.  It is already late for 6.5, but we try to make it for 6.6

**Comment by cbismuth** [comment]
No worries at all @mayya-sharipova, I'm sure you're hard working on many other things. I'll check the build failure and let you know, thanks :wink:

**Comment by MirAkram** [comment]
LG-V10

**Comment by cbismuth** [comment]
After rebasing with `master`, the `buildBwcVersion` Gradle task is back to green on my side.

Could you please rerun test? Thanks :+1:

**Comment by cbuescher** [comment]
@elasticmachine test this please

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java [hunk 6]
```diff
@@ -83,15 +82,22 @@
...
-        //we should not set boost and query name for queries that don't parse it
+        QB query;
         if (supportsBoostAndQueryName()) {
-            if (randomBoolean()) {
-                query.boost(2.0f / randomIntBetween(1, 20));
-            }
-            if (randomBoolean()) {
-                query.queryName(createUniqueRandomName());
-            }
+            query = createTestQueryBuilder(true, true);
```
Comment (mayya-sharipova) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java]: May be instead  of `supportsBoostAndQueryName()` create two methods `supportsBoost()` and `supportsQueryName()` that others can override?

File: server/src/test/java/org/elasticsearch/index/query/SpanTermQueryBuilderTests.java [hunk 7]
```diff
@@ -80,16 +80,20 @@ protected void doAssertLuceneQuery(SpanTermQueryBuilder queryBuilder, Query quer
      * @return an array of random {@link SpanTermQueryBuilder} with same field name
      */
     public SpanTermQueryBuilder[] createSpanTermQueryBuilders(int amount) {
+        return createSpanTermQueryBuilders(amount, true, true);
+    }
+
+    public SpanTermQueryBuilder[] createSpanTermQueryBuilders(int amount, boolean supportsBoost, boolean supportsQueryName) {
```
Comment (mayya-sharipova) [inline server/src/test/java/org/elasticsearch/index/query/SpanTermQueryBuilderTests.java]: I have noticed that you always call `createSpanTermQueryBuilders(_, false, true)` with `supportsBoost=false` and `supportQueryName=true`.  I don't think we need these parameters at all.  Looks like this method is only called from complex span queries, so you can just assume that we don't set boosting here, and just remove  the lines where the boost is set:
```
     spanTermQuery.boost(2.0f / randomIntBetween(1, 20));
```

**Review by mayya-sharipova — CHANGES_REQUESTED** [review]
@cbismuth  Thanks for continuing working on this. Overall, the PR is looking great and almost done! 

I suggested small restructuring to make the code cleaner.  Another thing is that we need to add some documentation that we don't support boosting on inner span queries (e.g. we can add this to span-queries.asciidoc).

File: server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java [hunk 8]
```diff
@@ -167,6 +167,17 @@ public final QB boost(float boost) {
         return (QB) this;
     }
 
+    static void checkNoBoost(String queryName, String fieldName, XContentParser parser, SpanQueryBuilder clause) {
```
[the remark this answers is not in the document]
Comment (cbismuth) [reply server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: * Static methods in an interface can only be called from the enclosing interface, so we can't move this method in `SpanQueryBuilder` and keep it static as it's called elsewhere.
* Making `SpanQueryBuilder` an abstract class isn't an option either as classes which implement it already extend other classes.
* Making `checkNoBoost` a default method in the `SpanQueryBuilder` isn't an option as default methods can't be called from static contexts, and callers need it to be static static to be registered in `SearchModule#registerQueryParsers` (see [here](https://github.com/elastic/elasticsearch/blob/e28509fbfea45bd381708a962ab4bdc04dda1a47/server/src/main/java/org/elasticsearch/search/SearchModule.java#L747)).

The option I've chosen is to add a package-private inner class `SpanQueryBuilderUtil` in the `SpanQueryBuilder` interface, see commit 9c751b10257ee79741f2c6fcbadb15271d9ee8f3.

File: server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java [hunk 9]
```diff
@@ -167,6 +167,17 @@ public final QB boost(float boost) {
...
+        try {
+            if (clause.boost() != AbstractQueryBuilder.DEFAULT_BOOST) {
+                throw new ParsingException(parser.getTokenLocation(),
+                    queryName + " [" + fieldName + "] can't have non-default boost value [" + clause.boost() + "]");
+            }
```
[the remark this answers is not in the document]
Comment (cbismuth) [reply server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: Totally, fixed in 8a7574704b5870a363d5b5c36714be2cc7607385.

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java (hunk 6 above)
Comment (cbismuth) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java]: Yes! I'm glad you ask for it, that's what I did first before reverting my change to no cause too much changes. Refactoring is back again in 8e9c70d1ebcf6a277024eed8efd8ed180c76c776 :+1:

File: server/src/test/java/org/elasticsearch/index/query/SpanTermQueryBuilderTests.java (hunk 7 above)
Comment (cbismuth) [reply server/src/test/java/org/elasticsearch/index/query/SpanTermQueryBuilderTests.java]: The `span_first` query supports boost and we use the boost and query name [here](https://github.com/elastic/elasticsearch/pull/34112/files/0ac11ecd4fda4feab10d13f0b1595e35d8f79df0#diff-fd26d71bc739f4672fae13a9d0a32a56R83).

Query name is always `true`, so we can remove this parameter, here is a commit to do so f2861cdae54126cac275d3eaf86b72acd47cd376.

File: server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: @cbismuth  thanks using `SpanQueryBuilderUtil` seems to be a very good solution

File: server/src/test/java/org/elasticsearch/index/query/SpanTermQueryBuilderTests.java (hunk 7 above)
Comment (mayya-sharipova) [reply server/src/test/java/org/elasticsearch/index/query/SpanTermQueryBuilderTests.java]: @cbismuth  can you please explain why `span_first` has a special treatment?  Shouldn't we disallow for its single inner span query to have `boost` as well?

File: server/src/test/java/org/elasticsearch/index/query/SpanContainingQueryBuilderTests.java [hunk 10]
```diff
@@ -89,5 +91,92 @@ public void testFromJson() throws IOException {
 
         assertEquals(json, 2, ((SpanNearQueryBuilder) parsed.bigQuery()).clauses().size());
         assertEquals(json, "foo", ((SpanTermQueryBuilder) parsed.littleQuery()).value());
+        assertEquals(json, 2.0, parsed.boost(), 0.0);
+    }
+
+    public void testFromJson_withNonDefaultBoost_inBigQuery() {
```
Comment (mayya-sharipova) [inline server/src/test/java/org/elasticsearch/index/query/SpanContainingQueryBuilderTests.java]: can we NOT use "_" here, just a camel case!

File: server/src/test/java/org/elasticsearch/index/query/SpanContainingQueryBuilderTests.java [hunk 11]
```diff
@@ -89,5 +91,92 @@ public void testFromJson() throws IOException {
...
+                "    \"boost\" : 1.0\n" +
+                "  }\n" +
+                "}";
+
+        Exception exception = expectThrows(ParsingException.class, () -> parseQuery(json));
+        assertThat(exception.getMessage(),
+            equalTo("span_containing [big] as a nested span clause can't have non-default boost value [2.0]"));
+    }
+
+    public void testFromJson_withNonDefaultBoost_inLittleQuery() {
```
Comment (mayya-sharipova) [inline server/src/test/java/org/elasticsearch/index/query/SpanContainingQueryBuilderTests.java]: the same thing just use camel case, without "_"

**Review by mayya-sharipova — CHANGES_REQUESTED** [review]
@cbismuth  Great changes so far! Thanks a lot for your persistence!
Just a couple of things to do:
1) Please explain why `span_first` needs  a special treatment, and if it is not we need to check boost for its inner query as well, and `createSpanTermQueryBuilders` doesn't need to have a `boost` parameter
2) remove all underscores in function names

File: server/src/test/java/org/elasticsearch/index/query/SpanContainingQueryBuilderTests.java (hunk 10 above)
Comment (cbismuth) [reply server/src/test/java/org/elasticsearch/index/query/SpanContainingQueryBuilderTests.java]: Sure, fixed in 043020b786ec289d20a6d72b3cf50af37831fb37.

File: server/src/test/java/org/elasticsearch/index/query/SpanContainingQueryBuilderTests.java (hunk 11 above)
Comment (cbismuth) [reply server/src/test/java/org/elasticsearch/index/query/SpanContainingQueryBuilderTests.java]: Yes, part of 043020b786ec289d20a6d72b3cf50af37831fb37.

**Comment by cbismuth** [comment]
Hi @mayya-sharipova, my bad `span_first` queries are compound span queries with a `match` inner query, therefore they should not support boost. I've removed support from them in c96aaa9aa5c25f0473f7e702a799034557ba06d1 (and then unnecessary `supportsBoost` parameter has been remove from test API in 11f6060442ac10c0c736ad16055143a162747949).

Besides, all `_` characters have been remove from test names in 043020b786ec289d20a6d72b3cf50af37831fb37.

Thank you!

**Review by mayya-sharipova — APPROVED** [review]
@cbismuth  Great changes! I approve the PR as the code and the tests are great! But we need to add two documentation  changes before we merge a PR:

1) Add a sentence into `span-queries.asciidoc` that we don't support boosting on inner span queries

2) Add an item in [breaking changes](https://github.com/elastic/elasticsearch/blob/master/docs/reference/migration/migrate_7_0/search.asciidoc)  that setting a boost is forbidden on inner nested span queries

**Comment by cbismuth** [comment]
Thanks a lot @mayya-sharipova. What do you think of this disclaimer below (adapted from [the explanation](https://discuss.elastic.co/t/troubles-with-complex-span-query-term-boosting/116106/2) of @jpountz).

> It is only allowed to set `boost` on an outer span query. Compound span queries, like `span_near`, only use the list of matching spans of sub span queries in order to find their own spans, which they then use to produce a score. Scores are never computed on sub span queries, which is the reason why boosts are not allowed: they only influence the way scores are computed, not spans.

File: server/src/test/java/org/elasticsearch/index/query/SpanTermQueryBuilderTests.java (hunk 7 above)
Comment (cbismuth) [reply server/src/test/java/org/elasticsearch/index/query/SpanTermQueryBuilderTests.java]: Totally. It has been removed in c96aaa9aa5c25f0473f7e702a799034557ba06d1.

File: server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (cbismuth) [reply server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: Thank you @mayya-sharipova :+1:

File: docs/reference/migration/migrate_7_0/search.asciidoc [hunk 12]
```diff
@@ -18,12 +18,14 @@
...
 
 *   Attempts to generate multi-term phrase queries against non-text fields
-    with a custom analyzer will now throw an exception
+    with a custom analyzer will now throw an exception.
 
 *   An `envelope` crossing the dateline in a `geo_shape `query is now processed
     correctly when specified using REST API instead of having its left and
     right corners flipped.
 
+*   Attempts to set `boost` on sub span queries will now throw an parsing exception.
```
Comment (mayya-sharipova) [inline docs/reference/migration/migrate_7_0/search.asciidoc]: " sub span queries" - > inner span queries

"an parsing exception" - > "a parsing exception"

**Comment by mayya-sharipova** [comment]
@cbismuth  I liked the test for  `span-queries.asciidoc` file, the only thing I would change: instead of "sub span queries" -> "inner span queries"

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

**Comment by mayya-sharipova** [comment]
Tests failed with unrelated failures, we can try again later

**Comment by cbismuth** [comment]
I've rebased with `master` it will probably help `bwc` tests to pass.

**Comment by cbismuth** [comment]
Thanks @mayya-sharipova, documentation fixes are pushed in 78e8c2154d29d2955550d736e26149f71ba16135.

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

**Comment by mayya-sharipova** [comment]
@cbismuth  sorry that it takes so long for this PR.  Do you mind to address documentation changes I asked for and resolve a conflict with `docs/reference/migration/migrate_7_0/search.asciidoc`. I will try to run a build later today as well, so we can make a build pass for this PR.

**Comment by cbismuth** [comment]
No worries @mayya-sharipova, conflict resolved, Java API documentation reverted in 1fb94ee26e6882e59c30c913ff9777c84b582bb0 and line length fixed in 94b5bcb19d36cafaf543b082901186ed5b5db544, thank you.

**Comment by mayya-sharipova** [comment]
@cbismuth  you are golden!

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

**Comment by cbismuth** [comment]
As far as I can see in the Jenkins output logs, the two `bwc` build commands are those below. I'll check these on my side and let you know as soon as they are green on my laptop :+1:

```bash
./gradlew --info --stacktrace --parallel --scan -Porg.elasticsearch.acceptScanTOS=true -Dscan.tag.CI clean
./gradlew --info --parallel --scan -Porg.elasticsearch.acceptScanTOS=true -Dscan.tag.CI precommit buildBwc
```

**Comment by cbismuth** [comment]
These two tasks are green on my side, let's give it a another try :+1:

**Comment by cbuescher** [comment]
@cbismuth @mayya-sharipova just passing by, but I can re-trigger CI
@elasticsearch test this please

**Comment by cbuescher** [comment]
@cbismuth lets try, thanks for monitoring the PR. If remote call issue you see repeats I probably need to dif a bit deeper within our infra. The other failure in ci-1 is something I just saw on another branch failing, but lets hope its a rare one (just for the sake of this PR, I'll try to repro that one)
@elasticmachine test this please

**Comment by cbismuth** [comment]
@cbuescher, unfortunately, still no luck.

```
11:49:59 Caused: java.io.IOException: Remote call on JNLP4-connect connection from static.149.132.251.148.clients.your-server.de/148.251.132.149:52722 failed
```

**Comment by cbuescher** [comment]
@cbismuth kk will watch out if this happens in our CI atm in general and ping infra about it

**Comment by cbismuth** [comment]
Great, thank you @cbuescher :+1:

**Comment by cbuescher** [comment]
There seems to have issue with CI this morning, but they should be resolved I was told and this worker seems to look okay from a first glance. Lets try again...
@elasticmachine test this please

**Comment by cbismuth** [comment]
Same issue :'(

```
12:45:00 Caused: java.io.IOException: Remote call on JNLP4-connect connection from static.149.132.251.148.clients.your-server.de/148.251.132.149:52722 failed
```

**Comment by fatmcgav** [comment]
As per discussion with @cbuescher on [Slack](https://elastic.slack.com/archives/C0D92ER9P/p1542974785581800), I've restarted the `jenkins-swarm` on [`worker-844165`](https://elasticsearch-ci.elastic.co/computer/worker-844165).

That should resolve the issues with this worker...

**Comment by cbismuth** [comment]
The build is green, thanks a lot @cbuescher and @fatmcgav :+1:

@mayya-sharipova, this PR is ready for merge :smile:

**Comment by mayya-sharipova** [comment]
@cbismuth  Merged - congratulations! I will later backport the changes to 6.6 as well

**Comment by cbismuth** [comment]
That is really great :smile: thanks a lot @mayya-sharipova for your help :+1:

**Comment by mayya-sharipova** [comment]
PR for backport: https://github.com/elastic/elasticsearch/pull/35967.  PR doesn't need review, just to make sure the build passes. Once the build is passed, we can merge

**Pull request merged**